include/fiction/technology/charge_distribution_surface.hpp

```diff
@@ -40,75 +45,720 @@ class charge_distribution_surface<Lyt, false> : public Lyt
+    }
+
+    /**
+     * Check if any SiDB is eqaul the given charge state.
+     *
+     * @param cs charge state for which it is checked whether any SiDB has it.
+     */
+    [[nodiscard]] bool charge_exists(const sidb_charge_state& cs) const
+    {
+        return std::accumulate(strg->cell_charge.begin(), strg->cell_charge.end(), true,
+                               [&cs](bool acc, const sidb_charge_state& c) { return acc && c == cs; });
+    }
+
+    /**
+     * Retrieves the physical parameters of the simulation.
+     *
+     * @return physical_params struct containing the physical parameters of the simulation.
+     */
+    [[nodiscard]] physical_params get_phys_params() const
+    {
+        return strg->phys_params;
+    }
+
+    /**
+     * This function assigns the given charge state to the cell of the layout at the specified index. It updates the
+     `cell_charge` member of `strg` object with the new charge state of the specified cell.
+     *
+     * @param i The index of the cell.
+     * @param cs The charge state to be assigned to the cell.
+     */
+    void state_index(const uint64_t& i, const sidb_charge_state& cs) const
+    {
+        strg->cell_charge[i] = cs;
+    }
+
+    /**
+     * This function assigns the given charge state to the given cell of the layout.
+     *
+     * @param cell The cell to which a state of charge is to be assigned.
+     * @param cs The charge state to be assigned to the cell.
+     */
+    void assign_charge_state_cell(const typename Lyt::cell& cell, const sidb_charge_state& cs) const
+    {
+        auto index = cell_to_index(cell);
+        if (index != -1)
+        {
+            strg->cell_charge[static_cast<uint64_t>(index)] = cs;
+        }
+        else
+        {
+            std::cout << "not part of the charge distribution surface" << std::endl;
+        }
+    }
+
+    /**
+     * This function assigns the given charge state to the cell (accessed by the input index) of the layout.
+     *
+     * @param index The index of the cell to which a state of charge is to be assigned.
+     * @param cs The charge state to be assigned to the cell.
+     */
+    void assign_charge_state_index(const uint64_t& index, const sidb_charge_state& cs) const
+    {
+        strg->cell_charge[index] = cs;
+    }
+
+    /**
+     * Sets the charge state of all SiDBs in the layout to a given charge state.
+     *
+     * @param cs The charge state to be assigned to all the SiDBs.
+     */
+    void set_charge_states(const sidb_charge_state& cs) const
+    {
+        for (uint64_t i = 0u; i < strg->cell_charge.size(); i++)
+        {
+            strg->cell_charge[i] = cs;
+        }
+    }
+
+    /**
+     * Returns the charge state of a cell of the layout at a given index.
+     *
+     * @param index The index of the cell.
+     * @return The charge state of the cell at the given index.
+     */
+    [[nodiscard]] sidb_charge_state get_charge_state_index(const uint64_t& index) const noexcept
```

Comment:
`get_charge_state_by_index`

otherwise, it is assumed that you return the index